demonstrations/tutorial_adaptive_circuits.py

```diff
@@ -0,0 +1,303 @@
+
+##############################################################################
+# Similarly to the double excitation gates, we select those single excitations that have a gradient
+# larger than a predefined threshold.
+
+singles_select = [singles[i] for i in range(len(singles)) if abs(grads[i]) > 1.0e-5]
+singles_select
+
+##############################################################################
+# We now have all of the gates we need to build our circuit and perform one final step of
+
+# optimization to get the ground-state energy. The resulting energy must be the exact energy of the
+# ground electronic state of LiH which is -7.8825378193 Ha.
+
+cost_fn = qml.ExpvalCost(circuit_1, H, dev, optimize=True)
+
+params = [0.0] * len(doubles_select + singles_select)
+
+for n in range(20):
+    t1 = time.time()
+    params, energy = opt.step_and_cost(cost_fn, params, excitations=doubles_select + singles_select)
+    t2 = time.time()
+    print("n = {:},  E = {:.8f} H, t = {:.2f} s".format(n, energy, t2 - t1))
+
+##############################################################################
+# Success! We obtained the exact ground state energy of LiH, within chemical accuracy, by having
+# only 10 gates in our circuit. This is less than half of the total number of single and double
+# excitations of LiH (24).
+
+##############################################################################
+# Sparse Hamiltonians
+# -------------
+#
+# Molecular Hamiltonians and quantum states are sparse. For instance, let’s look at the Hamiltonian
+# we built for LiH. We can compute its matrix representation in the computational basis using the
+# function :func:`~.pennylane.utils.sparse_hamiltonian` of PennyLane. This function
+# returns the matrix in the SciPy `sparse coordinate <https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html>`_ format.
+
+H_sparse = qml.utils.sparse_hamiltonian(H)
+H_sparse
+
+##############################################################################
+# The matrix has :math:`1024^2=1,048,576` entries, but only 11264 of them
+# are non-zero. Leveraging this sparsity in the simulations can significantly reduce the
+# simulation times. We use the implemented functionality in PennyLane for computing the expectation
+# value of the sparse Hamiltonian observable. This can reduce the cost of simulations by
+# orders of magnitude depending on the molecular size. We use the selected gates in the previous
+# steps and perform the final optimization step with the sparse method.
+
+dev = qml.device("default.qubit", wires=qubits)
+opt = qml.GradientDescentOptimizer(stepsize=0.5)
+
+excitations = doubles_select + singles_select
+params = [0.0] * len(excitations)
+
+
+@qml.qnode(dev, diff_method="parameter-shift")
+def circuit(params):
+    hf_state = qchem.hf_state(2, qubits)
+    qml.BasisState(np.array(hf_state, requires_grad=False), wires=range(qubits))

```

Comment:
```suggestion
    qml.BasisState(np.array(hf_state, requires_grad=False), wires=range(qubits))
```
Not sure this needs the `np.array` here either.